### PR TITLE
Refactoring to multi-process approach

### PR DESCRIPTION
**Pull request by avoidwork**
screen shot 2017-11-12 at 3 40 06 pm

- [x] Spins up a ps per cache to minimize JIT state pollution (gc, opt/deopt, etc.)
- [x] Updated README content
- [x] Added missing `should` module to deal with the npm bug

### CODE REVIEW

File: worker.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    'hyperlru-object': max => hyperlruObject({max}),
+    'hyperlru-map': max => hyperlruMap({max}),
+    lru_cache: n => new LRUCache(n),
+    lru: require('lru'),
+    mkc: max => new MKC({max})
+  },
+  num = 2e5,
+  evicts = num * 2,
+  times = 5,
+  x = 1e6,
```
Comment (Kikobeats) [inline worker.js]: what is that?

Comment (avoidwork) [reply worker.js]: i don't understand the question, it's a number; it's used to turn nanoseconds into milliseconds.

**Comment by avoidwork** [comment]
Yes it's necessary, all it does is setup the vm context and decorate a few functions so the child process mimics a web worker.

File: worker.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    for (let i = 0; i < num; i++) lru.set(i, Math.random());
+    time.update.push(utimer.stop().diff() / x);
+
+    const g2timer = precise().start();
+    for (let i = 0; i < num; i++) lru.get(i);
+    time.get2.push(g2timer.stop().diff() / x);
+
+    let etimer = precise().start();
+    for (let i = num; i < evicts; i++) lru.set(i, Math.random());
+    time.evict.push(etimer.stop().diff() / x);
```
Comment (avoidwork) [inline worker.js]: `x` is moving the decimal on the cpu time (ns); the original README was "ops/ms", this is the most efficient way to get at it.

**Comment by avoidwork** [comment]
Merged since there didn't seem to be an objection.

**Comment by avoidwork** [comment]
btw, i tried compressing this code further but you run into diminishing returns with the JIT... too many arrays and perf goes to shit.

**Comment by avoidwork** [comment]
i7_7700k

ran a fresh clone on my i7-7700k ... i think it's working :D

**Comment by Kikobeats** [comment]
@avoidwork

What do you think about now results are totally different? I mean, could be possible determinate why libraries are faster than other?

For example, looks like use prototype inheritance is a must for be faster.

**Comment by avoidwork** [comment]
from what i looked at the faster ones are striving for least work possible (mine fails in a few spots). i don't think the prototype does much in this case, unless a lot of your core methods are in an outer scope.

mine can/should be f'd up with megamorphic objects because it's not using a `Map` to hold things due to cost of serialization if you want to sync with another cache. ¯\_(ツ)_/¯

the benchmark itself plays a huge role in the overall ops/ms timing. if you take away that imperial `while()` and try to compress it via an array of attributes to test with a loop you cut the perf by 1/3 minimum.

**Comment by avoidwork** [comment]
@Kikobeats I just pushed up a revised benchmark; I noticed 2 mistakes yesterday...

1. the math was wrong, i had erroneously moved the decimal to calculate `ops/sec` instead of `ops/ms`

2. CPU contention created by running tests ASAP slowed down the faster caches, so now it's in an async pipeline such that the 'next' test runs after the 'prev' resolves, while still using a `Promise.all()` to collect all of the results

**Comment by Kikobeats** [comment]
so results are not different, we need to update them from readme.md

I'm thinking in setup travis just for run the script in each build and ouptut the benchmark; then we can paste the result into the readme.

**Comment by avoidwork** [comment]
Results are still different, the spread is wider by about 4x from og readme; it's already been updated btw.

The newest benchmark does run them as you're like; it's a simple trick I used last year for spectron tests on an electron app; you create a promise for each test and exec the 'next' test from the resolution of the 'prev' test, using the array of promises for collection.

This minimizes lighting up all CPU cores at once and creating a resource & scheduling problem.

**Comment by Kikobeats** [comment]
@avoidwork we have a little typo: in the bench, `js-lru` appears twice times.

for the rest looks awesome 💯

**Comment by avoidwork** [comment]
ugh, ill fix right now. sorry!

**Comment by avoidwork** [comment]
Updated readme! I also ran the benchmark on my i7 again & the results are more believable 😂 

better

**Pull request merged**